tsdb/engine/tsm1/tombstone.go

```diff
@@ -183,43 +182,52 @@ func (t *Tombstoner) Delete() error {
 
 // HasTombstones return true if there are any tombstone entries recorded.
 func (t *Tombstoner) HasTombstones() bool {
-	files := t.TombstoneFiles()
+	stats := t.TombstoneStats()
+	if !stats.TombstoneExists {
+		return false
+	}
+	if stats.Size > 0 {
+		return true
+	}
+
 	t.mu.RLock()
 	n := len(t.tombstones)
 	t.mu.RUnlock()
 
-	return len(files) > 0 && files[0].Size > 0 || n > 0
+	return n > 0
 }
 
 // TombstoneFiles returns any tombstone files associated with Tombstoner's TSM file.
-func (t *Tombstoner) TombstoneFiles() []FileStat {
+func (t *Tombstoner) TombstoneStats() TombstoneStat {
 	t.mu.RLock()
 	if t.statsLoaded {
-		stats := t.fileStats
+		stats := t.tombstoneStats
 		t.mu.RUnlock()
 		return stats
 	}
 	t.mu.RUnlock()
 
 	stat, err := os.Stat(t.tombstonePath())
-	if os.IsNotExist(err) || err != nil {
+	if err != nil {
 		t.mu.Lock()
 		// The file doesn't exist so record that we tried to load it so
 		// we don't continue to keep trying.  This is the common case.
 		t.statsLoaded = os.IsNotExist(err)
-		t.fileStats = t.fileStats[:0]
+		t.tombstoneStats.TombstoneExists = false
+		stats := t.tombstoneStats
 		t.mu.Unlock()
-		return nil
+		return stats
```

Comment:
Are we always sure that t.tomstoneStats != nil

Reply:
Yes, since it's not tracked as a pointer any more (see declaration [here](https://github.com/influxdata/influxdb/pull/20909/files#diff-c196bc5a05dd81c2e1d38e91ab08cff51f2e6413dc091f92c364e376c1818ed2R44))

Reply:
In the zero case, the struct will be returned with `TombstoneExists: false`. All the callsites have been updated to check that field before using the rest of the returned data.